File: ios/telobike/telobike/TBStationTableViewCell.m
//
//  TBStationTableViewCell.m
//  telobike
//
//  Created by Elad Ben-Israel on 9/27/13.
//  Copyright (c) 2013 Elad Ben-Israel. All rights reserved.
//

@import QuartzCore;

#import <QuartzCore/QuartzCore.h>
#import "TBStationTableViewCell.h"
#import "TBAvailabilityView.h"
#import "TBTintedView.h"

@interface TBStationTableViewCell () <CLLocationManagerDelegate>

@property (strong, nonatomic) IBOutlet UILabel* stationNameLabel;
@property (strong, nonatomic) IBOutlet UILabel* subtitleLabel;
@property (strong, nonatomic) IBOutlet TBAvailabilityView* availabilityView;
@property (strong, nonatomic) IBOutlet TBTintedView* availabilityIndicatorView;
@property (strong, nonatomic) CLLocationManager* locationManager;
@property (strong, nonatomic) MKDistanceFormatter* distanceFormatter;

@end

@implementation TBStationTableViewCell

- (void)awakeFromNib {
    self.locationManager = [[CLLocationManager alloc] init];
    [self.locationManager startUpdatingLocation];
    self.locationManager.delegate = self;
    
    self.distanceFormatter = [[MKDistanceFormatter alloc] init];
    self.distanceFormatter.units = MKDistanceFormatterUnitsMetric;
    self.distanceFormatter.unitStyle = MKDistanceFormatterUnitStyleAbbreviated;
    
    
    // add a hairline layer as a separator
    CALayer* separator = [[CALayer alloc] init];
    separator.frame = CGRectMake(0.0f, 0.0f, self.bounds.size.width, 0.5f);
    separator.backgroundColor = [[UIColor colorWithWhite:0.8f alpha:1.0f] CGColor];
    [self.layer addSublayer:separator];
}

- (void)setStation:(TBStation *)station {
    _station = station;
    
    self.availabilityView.station = self.station;
    self.stationNameLabel.text = self.station.stationName;
    self.availabilityIndicatorView.fillColor = self.station.indicatorColor;
    
    [self locationManager:self.locationManager didUpdateLocations:nil];
}

- (void)prepareForReuse {
    [super prepareForReuse];
    self.subtitleLabel.hidden = YES;
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations {
    
    // hide label if no location services
    if ([CLLocationManager authorizationStatus] != kCLAuthorizationStatusAuthorized) {
        self.subtitleLabel.hidden = YES;
    }
    
    if (self.locationManager.location) {
        CLLocationDistance distance = [self.station distanceFromLocation:self.locationManager.location];
        NSMutableAttributedString* desc = [[NSMutableAttributedString alloc] init];

        if (distance < 100000) {
            [desc appendAttributedString:[[NSAttributedString alloc] initWithString:[self.distanceFormatter stringFromDistance:distance] attributes:@{ NSForegroundColorAttributeName: [UIColor lightGrayColor] }]];
        }
        else {
            [desc appendAttributedString:[[NSAttributedString alloc] initWithString:NSLocalizedString(@"far", nil) attributes:@{ NSForegroundColorAttributeName: [UIColor lightGrayColor] }]];
        }
        
        self.subtitleLabel.attributedText = desc;
        self.subtitleLabel.hidden = NO;
    }
    else {
        self.subtitleLabel.hidden = YES;
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    // hide when no location
    self.subtitleLabel.hidden = YES;
}

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    if (status != kCLAuthorizationStatusAuthorized) {
        self.subtitleLabel.hidden = YES;
    }
}

@end

File: ios/telobike/telobike/TBStationTableViewCell.swift
//
//  TBStationTableViewCell.swift
//  telobike
//
//  Created by Amit Attias on 10/15/14.
//  Copyright (c) 2014 Elad Ben-Israel. All rights reserved.
//

import UIKit
import CoreLocation

let STATION_CELL_REUSE_IDENTIFIER: String = "STATION_CELL"

class TBStationTableViewCell: UITableViewCell, CLLocationManagerDelegate  {

    // public
    var station: TBStation! {
        didSet {
            self.availabilityView.station = station
            self.stationNameLabel.text = station.stationName
            self.availabilityIndicatorView.fillColor = station.indicatorColor
            self.locationManager(self.locManager, didUpdateLocations: [])
        }
    }

    // private
    @IBOutlet private var availabilityIndicatorView: TBTintedView!
    @IBOutlet private var availabilityView: TBAvailabilityView!
    @IBOutlet private var subtitleLabel: UILabel!
    @IBOutlet private var stationNameLabel: UILabel!
    private var locManager: CLLocationManager!
    private var distanceFormatter: MKDistanceFormatter!

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        
        self.locManager = CLLocationManager()
        if #available(iOS 8.0, *) {
            self.locManager.requestWhenInUseAuthorization()
        }
        self.locManager.startUpdatingLocation()
        self.locManager.delegate = self;
        
        self.distanceFormatter = MKDistanceFormatter()
        
        self.distanceFormatter.units = .Metric;
        self.distanceFormatter.unitStyle = .Abbreviated;
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        self.subtitleLabel.hidden = true
    }

    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // hide label if no location services
        if (CLLocationManager.authorizationStatus() != .Authorized) {
            self.subtitleLabel.hidden = true
        }
        
        if let location = self.locManager.location {
            let distance = self.station.location.distanceFromLocation(location)
            if (distance < 100_000) {
                self.subtitleLabel.text = self.distanceFormatter.stringFromDistance(distance)
            }
            else {
                self.subtitleLabel.attributedText = nil
                #if DEBUG
                    self.subtitleLabel.text = self.distanceFormatter.stringFromDistance(distance)
                #endif
            }
            
            self.subtitleLabel.hidden = false
        }
        else {
            self.subtitleLabel.hidden = true
        }

    }

    func locationManager(manager: CLLocationManager, didFailWithError error: NSError) {
        // hide when no location
        self.subtitleLabel.hidden = true
    }
    
    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
        if (status != .Authorized) {
            self.subtitleLabel.hidden = true
        }
    }
}